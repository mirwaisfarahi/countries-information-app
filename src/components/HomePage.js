import React, { useEffect, useState, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchCountries } from '../redux/country/country';
import { fetchCountrySpotlight } from '../redux/countrySpotlight/countrySpotlight';
import Country from './Country';
import '../App.css';
import '../index.css';

const HomePage = () => {
  const { countrySpotlight } = useSelector((state) => state.countrySpotlight);
  const { countriesList } = useSelector((state) => state.countries);
  const countriesInputRef = useRef();
  const dispatch = useDispatch();

  const [filter, setFilter] = useState(countriesList);

  const filterCountries = (e) => {
    const target = e.target.value;
    if (target === 'all') {
      setFilter(countriesList);
    } else {
      setFilter(countriesList.filter((el) => el.region === target));
    }
  };

  const filterCountriesByName = () => {
    const searchValue = countriesInputRef.current.value;

    if (searchValue.trim()) {
      const fetchSearch = async () => {
        const req = await fetch(`https://restcountries.com/v3.1/name/${searchValue}`);
        const res = await req.json();
        setFilter(res);
      };
      try {
        fetchSearch();
      } catch (error) {
        console.log(error);
      }
    } else {
      setFilter(countriesList);
    }
  };

  useEffect(() => {
    dispatch(fetchCountrySpotlight());
    dispatch(fetchCountries());
    setFilter(null);
  }, [dispatch]);

  return (
    <div className="countries-container">
      {countrySpotlight.map((country) => (
        <div className="spotlight-card" key={country.id}>
          <div className="flag-spot-container">
            <img className="spot-flag" src={country.flags.svg} alt="Flag" />
          </div>
          <div className="spot-details">
            <h2 className="spot-name">{country.name.common}</h2>
            <h3 className="spot-capital">
              Capital City:
              {country.capital}
            </h3>
            <h3 className="spot-population">
              Population:
              {country.population}
            </h3>
          </div>
        </div>
      ))}
      <div className="population-stats">
        <input className="stats-search" type="text" ref={countriesInputRef} onChange={filterCountriesByName} />
        <select
          className="stats-select"
          name="drop-down"
          onChange={filterCountries}
        >
          <option value="all">All</option>
          <option value="Asia">Asia</option>
          <option value="Africa">Africa</option>
          <option value="Americas">America</option>
          <option value="Europe">Europe</option>
          <option value="Oceania">Oceania</option>
        </select>
      </div>
      <Country countries={filter || countriesList} />
    </div>
  );
};

export default HomePage;
