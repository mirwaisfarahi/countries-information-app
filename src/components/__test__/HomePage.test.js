import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import HomePage from '../HomePage';
import store from '../../redux/store/configureStore';
import '@testing-library/jest-dom';

it('Check if the component has changed', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <HomePage />
        ,
      </BrowserRouter>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
