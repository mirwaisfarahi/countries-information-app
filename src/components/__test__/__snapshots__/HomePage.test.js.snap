// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if the component has changed 1`] = `
Array [
  <div
    className="countries-container"
  >
    <div
      className="population-stats"
    >
      <h2
        className="stats-title"
      >
        STATS BY COUNTRY POPULATION
      </h2>
    </div>
    <div
      className="country-container"
    />
  </div>,
  ",",
]
`;
